File: frontend/src/app/utils/product-request.ts
import { Product } from '../models';

export type ProductRequestBody = Omit<Product, 'id' | 'categoryName' | 'savings' | 'onSale'>;

export function toProductRequest(
  source: Partial<Product> & { name: string; price: number; stock: number; active: boolean }
): ProductRequestBody {
  return {
    name: source.name,
    description: source.description ?? '',
    price: Number(source.price),
    originalPrice: source.originalPrice ?? Number(source.price),
    discountPercent: source.discountPercent ?? 0,
    stock: Number(source.stock),
    imageUrl: source.imageUrl ?? '',
    categoryId: source.categoryId,
    active: source.active !== false
  };
}

File: backend/src/main/java/com/ichrafsassi/ecommerce/dto/ProductRequest.java
package com.ichrafsassi.ecommerce.dto;

import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.Max;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;

import java.math.BigDecimal;

public record ProductRequest(
        @NotBlank String name,
        String description,
        @NotNull @DecimalMin("0.01") BigDecimal price,
        BigDecimal originalPrice,
        @Min(0) @Max(90) Integer discountPercent,
        @NotNull @Min(0) Integer stock,
        String imageUrl,
        Long categoryId,
        boolean active
) {}
